### ISSUE TITLE
problem with early instantiation of settings (due to `import alot` in hooks file)

### ISSUE THREAD

**Issue by lucc** [pazz/alot#1246]
Software Versions
-----------------

Python version: 3.6.5 (Arch Linux)

Notmuch version: 0.26.2 (Arch Linux)

Alot version (either a number or sha for master): master: 0.7-127-ge14488f1

Observed behavior
-----------------

I see an exception on the terminal when starting up alot. Functionality seems to be unobstructed. It runs just fine afterwards.

I ran `alot -c ~/alot.tmp -d debug -l logfile`
This is the output I see
```
ERROR:root:unable to load hooks file:/home/luc/.config/alot/hooks.py
Traceback (most recent call last):
  File "/home/luc/vcs/alot/alot/settings/manager.py", line 93, in read_config
    self.hooks = imp.load_source('hooks', hooks_path)
  File "/usr/lib/python3.6/imp.py", line 172, in load_source
    module = _load(spec)
  File "<frozen importlib._bootstrap>", line 684, in _load
  File "<frozen importlib._bootstrap>", line 665, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 678, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/home/luc/.config/alot/hooks.py", line 14, in <module>
    from alot.commands.envelope import SignCommand, EncryptCommand
  File "/home/luc/vcs/alot/alot/commands/__init__.py", line 12, in <module>
    from ..settings.const import settings
ImportError: cannot import name 'settings'
inside tmp hooks file
```
The file `~/alot.tmp` looks like this:
```
hooksfile = ~/tmp/config/alot/hooks.py
[accounts]
[[x]]
address = foo@bar
realname = lucas
sendmail_command = msmtp
sent_box = maildir:///home/luc/tmp/sent
draft_box = maildir:///home/luc/tmp/draft
```

The file `~/tmp/config/alot/hooks.py` looks like this:
```
print("inside tmp hooks file")
```
In the default hooks file (`~/.config/alot/hooks.py`) I have this line: `from alot.commands.envelope import SignCommand, EncryptCommand`. If I comment that out, I don't see the error.

The logfile doesn't have any info about that error!

Expected behavior
-----------------

No stack trace.

Additional Info
-----------------

The [`imp`](https://docs.python.org/3/library/imp.html) module is deprecated since 3.4. On StackOverflow they say one should use importlib: https://stackoverflow.com/questions/67631

**Comment by pazz** [comment]
Yes, as mentioned in #1055, I believe we should use https://stackoverflow.com/questions/67631/how-to-import-a-module-given-the-full-path/67692#67692

**Comment by lucc** [comment]
Notice that the first line of the error message mentions my default hooks file and the last line is from the tmp hooks file.

Also the offending `import` is only in the default hooks file!

**Comment by lucc** [comment]
With this diff on current master 
```diff
diff --git a/alot/settings/manager.py b/alot/settings/manager.py
index 56e6eec8..d6f3063d 100644
--- a/alot/settings/manager.py
+++ b/alot/settings/manager.py
@@ -90,6 +90,10 @@ class SettingsManager(object):
 
         hooks_path = os.path.expanduser(self._config.get('hooksfile'))
         try:
+            print('attempt to load hooks file: %s', hooks_path)
+            import traceback
+            for line in traceback.format_stack():
+                print(line.strip())
             self.hooks = imp.load_source('hooks', hooks_path)
         except:
             logging.exception('unable to load hooks file:%s', hooks_path)
```
and this invocation of alot: `alot -c ~/tmp/config/alot/config -l logfile -d debug`
I get this output on the terminal before alot fully starts up:
```
attempt to load hooks file: %s /home/luc/.config/alot/hooks.py
File "/home/luc/.local/bin/alot", line 11, in <module>
    load_entry_point('alot', 'console_scripts', 'alot')()
File "/usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 476, in load_entry_point
    return get_distribution(dist).load_entry_point(group, name)
File "/usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2700, in load_entry_point
    return ep.load()
File "/usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2318, in load
    return self.resolve()
File "/usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2324, in resolve
    module = __import__(self.module_name, fromlist=['__name__'], level=0)
File "<frozen importlib._bootstrap>", line 971, in _find_and_load
File "<frozen importlib._bootstrap>", line 955, in _find_and_load_unlocked
File "<frozen importlib._bootstrap>", line 665, in _load_unlocked
File "<frozen importlib._bootstrap_external>", line 678, in exec_module
File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
File "/home/luc/vcs/alot/alot/__main__.py", line 12, in <module>
    from alot.settings.const import settings
File "<frozen importlib._bootstrap>", line 971, in _find_and_load
File "<frozen importlib._bootstrap>", line 955, in _find_and_load_unlocked
File "<frozen importlib._bootstrap>", line 665, in _load_unlocked
File "<frozen importlib._bootstrap_external>", line 678, in exec_module
File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
File "/home/luc/vcs/alot/alot/settings/const.py", line 8, in <module>
    settings = SettingsManager()
File "/home/luc/vcs/alot/alot/settings/manager.py", line 56, in __init__
    self.reload()
File "/home/luc/vcs/alot/alot/settings/manager.py", line 71, in reload
    self.read_config()
File "/home/luc/vcs/alot/alot/settings/manager.py", line 95, in read_config
    for line in traceback.format_stack():
ERROR:root:unable to load hooks file:/home/luc/.config/alot/hooks.py
Traceback (most recent call last):
  File "/home/luc/vcs/alot/alot/settings/manager.py", line 97, in read_config
    self.hooks = imp.load_source('hooks', hooks_path)
  File "/usr/lib/python3.6/imp.py", line 172, in load_source
    module = _load(spec)
  File "<frozen importlib._bootstrap>", line 684, in _load
  File "<frozen importlib._bootstrap>", line 665, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 678, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/home/luc/.config/alot/hooks.py", line 14, in <module>
    from alot.commands.envelope import SignCommand, EncryptCommand
  File "/home/luc/vcs/alot/alot/commands/__init__.py", line 12, in <module>
    from ..settings.const import settings
ImportError: cannot import name 'settings'
attempt to load hooks file: %s /home/luc/tmp/config/alot/hooks.py
File "/home/luc/.local/bin/alot", line 11, in <module>
    load_entry_point('alot', 'console_scripts', 'alot')()
File "/home/luc/vcs/alot/alot/__main__.py", line 107, in main
    settings.read_config()
File "/home/luc/vcs/alot/alot/settings/manager.py", line 95, in read_config
    for line in traceback.format_stack():
inside tmp hooks file
```
Notice that it first tries to load the default hooks file. inside that one is the offending recursive import. Only after that failed with an ImportError it tries to load the tmp hooks file (and succeeds, see last line).

**Comment by pazz** [comment]
Have you tried commenting out [this line](https://github.com/pazz/alot/blob/master/alot/settings/manager.py#L56) ? It seems that the main function triggers the instanciation of a `SettingsManager` object *before* the commandline/config options are parsed. This shouldn't happen I think.

**Comment by lucc** [comment]
No I just get a different error and alot doesn't start.
```
attempt to load hooks file: %s /home/luc/tmp/config/alot/hooks.py
File "/home/luc/.local/bin/alot", line 11, in <module>
    load_entry_point('alot', 'console_scripts', 'alot')()
File "/home/luc/vcs/alot/alot/__main__.py", line 107, in main
    settings.read_config()
File "/home/luc/vcs/alot/alot/settings/manager.py", line 95, in read_config
    for line in traceback.format_stack():
inside tmp hooks file
Traceback (most recent call last):
  File "/home/luc/.local/bin/alot", line 11, in <module>
    load_entry_point('alot', 'console_scripts', 'alot')()
  File "/home/luc/vcs/alot/alot/__main__.py", line 107, in main
    settings.read_config()
  File "/home/luc/vcs/alot/alot/settings/manager.py", line 103, in read_config
    self._bindings.merge(newbindings)
AttributeError: 'NoneType' object has no attribute 'merge'
```

**Comment by pazz** [comment]
OK.

I think the issue with import orders leading to your problem with two
hook files is a separate issue from my original intent to replace `imp`.

Perhaps the way to go there is to replace the
`SettingsManager.read_config` logic in a way that it directly gets the
file path as a parameter. Our main routine `alot/__main__.py` can then
import the global settings instance (which instantiates it at this
point) and then call its `read_config` mathod with the right file path
(from argparse/defaults).

A nice side effect would be that this means fewer workarounds when
testing classes that import settings but where the tests don't rely on
valid user settings. IIRC, we needed to Mock around there.

**Comment by lucc** [comment]
I approved #1247 and renamed this issue.

We could also *not* treat the settings as a global object that will be imported by all other code but rather as an object that is created by the main routine and passed around as an additional attribute of the ui object or function argument.  But that might be a very big refactor.

But at the moment I think your solution is better.

**Comment by lucc** [comment]
Was not fixed by #1246 but hopefully will be fixed by #1251.

**Comment by pazz** [comment]
#1251 fixed the issue for you @lucc right? (if so please close this issue)

**Comment by lucc** [comment]
sorry forgot to close it.
